File: website/src/types/catalog.ts
export interface Credential {
  id: string
  name: string
  priority: number
  enabled: boolean
  verification_status: 'pending' | 'verified' | 'failed'
  verified_at: string | null
}
export interface ProviderModel {
  enabled: boolean
  etag: string
  id: string
  upstream_name: string
}
export interface Connection {
  id: string
  name: string
  base_url: string
  protocol: 'openai_chat' | 'openai_responses'
  credentials: Credential[]
  provider_models: ProviderModel[]
}
export interface Provider {
  id: string
  name: string
  connections: Connection[]
}
export interface Model {
  id: string
  name: string
  status: 'active' | 'disabled' | 'archived'
  names: { name: string; is_current: boolean; expires_at: string | null }[]
  bindings: {
    id: string
    provider_model_id: string
    provider_id: string
    connection_id: string
    upstream_name: string
    protocol: string
    weight: number
    ready: boolean
  }[]
  granted_user_ids: string[]
}
export interface CallableModel {
  id: string
  name: string
  status: Model['status']
  protocol: string
  protocols?: string[]
}
export interface PersonalKey {
  id: string
  name: string
  prefix: string
  status: 'pending' | 'active' | 'disabled' | 'revoked'
  model_ids: string[]
  expires_at: string | null
  created_at: string
  replaces_key_id: string | null
  delivery_expires_at: string | null
}
export interface KeyDelivery {
  key: PersonalKey
  secret: string
}

File: internal/routex/entity/catalog.go
package entity

import "time"

const ProtocolOpenAIChat = "openai_chat"

// Provider is a stable upstream supplier; transport and secrets belong to connections.
type Provider struct {
	ID        string `gorm:"primaryKey;size:30"`
	Name      string `gorm:"size:100;not null"`
	CreatedAt time.Time
}

type ProviderConnection struct {
	ID         string `gorm:"primaryKey;size:30"`
	ProviderID string `gorm:"size:30;not null"`
	Name       string `gorm:"size:100;not null"`
	BaseURL    string `gorm:"size:2048;not null"`
	Protocol   string `gorm:"size:30;not null"`
	CreatedAt  time.Time
}

type ProviderCredential struct {
	ID                 string `gorm:"primaryKey;size:30"`
	ConnectionID       string `gorm:"size:30;not null"`
	Name               string `gorm:"size:100;not null"`
	Ciphertext         string `gorm:"type:text;not null"`
	Priority           int    `gorm:"not null"`
	Enabled            bool   `gorm:"not null"`
	VerificationStatus string `gorm:"size:20;not null"`
	VerifiedAt         *time.Time
	CreatedAt          time.Time
}

type ProviderModel struct {
	ID           string `gorm:"primaryKey;size:30"`
	ConnectionID string `gorm:"size:30;not null"`
	UpstreamName string `gorm:"size:255;not null"`
	Disabled     bool   `gorm:"not null;default:false"`
	ETag         string `gorm:"size:64;not null;default:0"`
	CreatedAt    time.Time
}

// CredentialModelAccess records actual discovery results for one credential.
type CredentialModelAccess struct {
	CredentialID    string `gorm:"primaryKey;size:30"`
	ProviderModelID string `gorm:"primaryKey;size:30"`
}

type Model struct {
	ID        string `gorm:"primaryKey;size:30"`
	Status    string `gorm:"size:20;not null"`
	CreatedAt time.Time
}

// CurrentModelID is nullable and unique: at most one current name per model.
// Retired names remain reserved globally after their compatibility deadline.
type ModelName struct {
	Name           string  `gorm:"primaryKey;size:128"`
	ModelID        string  `gorm:"size:30;not null"`
	CurrentModelID *string `gorm:"size:30;uniqueIndex"`
	ExpiresAt      *time.Time
	CreatedAt      time.Time
}

type ModelProviderBinding struct {
	ID              string `gorm:"primaryKey;size:30"`
	ModelID         string `gorm:"size:30;not null"`
	ProviderModelID string `gorm:"size:30;not null"`
	Weight          int    `gorm:"not null"`
	CreatedAt       time.Time
}

type UserModelGrant struct {
	UserID    string `gorm:"primaryKey;size:30"`
	ModelID   string `gorm:"primaryKey;size:30"`
	CreatedAt time.Time
}
